### ISSUE TITLE
Bidirectional text support

### ISSUE THREAD

**Issue by yzn-h** [emilk/egui#1016]
**Describe the solution you'd like**
I'd love to be able to write in RTL languages like Arabic and Japanese.

**Describe alternatives you've considered**
there are tools to reverse RTL languages letters like this for [Arabic](https://www.arabic-keyboard.org/photoshop-arabic) but in egui I cant write RTL letters

**Additional context**
- great talk [link](https://youtu.be/eCxqAr7LjW8?t=385) 
- Godot engine 4.0 has an excellent implementation

**Comment by yzn-h** [comment]
how Godot managed to do it:
1- https://godotengine.org/article/complex-text-layouts-progress-report-1
2 - https://godotengine.org/article/complex-text-layouts-progress-report-2
3 - https://godotengine.org/article/complex-text-layouts-progress-report-3

**Comment by emilk** [comment]
Implementing right-to-left is a lot of work (I know, I've done it before). I rather switch egui to use a crate that does it all for me. I'm not sure which though. I haven't looked at the alternatives in a long while, but I've been keeping my eye on https://github.com/dfrg/swash by @dfrg (https://github.com/dfrg/swash_demo is _very_ impressive).

**Comment by parasyte** [comment]
Related to #56

**Comment by Mohsen7s** [comment]
@emilk Did you make a decision on this ? Any ETA ?

**Comment by emilk** [comment]
No, I have other things higher on my priority list right now. What's needed first is a bit of research in different text rendering/layout libraries (I mentioned swash above, but there might be other ones to check out).

**Comment by ElhamAryanpur** [comment]
Hello!

Any plans for this?

**Comment by yzn-h** [comment]
can't  [cosmic text](https://github.com/pop-os/cosmic-text) be used for egui ?

**Comment by parasyte** [comment]
Yes, and cosmic text may help with rendering quality as mentioned in https://github.com/emilk/egui/issues/2639#issuecomment-1422287556

**Comment by ElhamAryanpur** [comment]
bump

Are there any plans so far?

**Comment by ElhamAryanpur** [comment]
Bump

**Comment by emilk** [comment]
There's no plans to implement this in egui right now (it is _a lot_ of work). I rather switch out the text layout and rendering in egui than work on it further. I am curious to investigate if [cosmic text](https://github.com/pop-os/cosmic-text) would be a good fit.

**Comment by ElhamAryanpur** [comment]
So far cosmic text's roadmap seems to be almost done. I was thinking of giving it a try and sort of maybe hack a fix into egui later on. Can I post updates here or...somewhere else?

**Comment by emilk** [comment]
That sounds great @ElhamAryanpur ! I suggest opening a new issue here specifically about Cosmic Text and then posting your updates there! On of the things I'm curious about is how Cosmic Text does its text rendering

**Comment by tliron** [comment]
Adding [Parley](https://github.com/linebender/parley) as an option.

Generally the whole Linebender stack seems to be a very nice effort to unify work on these challenging low-level text and font issues.

**Comment by awildergoose** [comment]
Any updates on this? Really need this for my current project.
